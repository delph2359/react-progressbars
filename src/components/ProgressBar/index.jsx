import React, { Component } from "react";
import PropTypes from "prop-types";
import "./index.scss";

export default class ProgressBar extends Component {
  static propTypes = {
    percentage: PropTypes.number,
    limit: PropTypes.number
  };

  static defaultProps = {
    percentage: 0,
    limit: 100
  };

  render() {
    let { percentage, limit } = this.props;
    if (percentage < 0) percentage = 0;
    return (
      <div className="progress-bar" data-testid="progress-bar">
        <div
          className={
            percentage > limit ? "progress-fill limit" : "progress-fill"
          }
          style={{ width: `${Math.min(percentage, 100)}%` }}
        ></div>
        <span>{percentage}%</span>
      </div>
    );
  }
}
