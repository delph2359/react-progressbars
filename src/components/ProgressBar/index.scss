.progress-bar {
  height: 30px;
  width: 350px;
  border-radius: 15px;
  border: 1px solid #444;
  margin: 2%;
  position: relative;
  span {
    position: absolute;
    z-index: 2;
    text-align: center;
    width: 100%;
  }
}

.progress-fill {
  height: 100%;
  background: #0093d1;
  border-radius: inherit;
  position: absolute;
  z-index: 1;
  transition: width 0.05s ease-in;
  &.limit {
    background: #f2635f;
  }
}
