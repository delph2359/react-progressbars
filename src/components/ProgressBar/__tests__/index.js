import "@testing-library/jest-dom/extend-expect";
import React from "react";
import { render } from "@testing-library/react";
import ProgressBar from "../";

describe("Progress bar", () => {
  test("should render basic Progress bar (no props)", () => {
    const { container, getByText } = render(<ProgressBar />);
    expect(container).toMatchSnapshot();
    expect(getByText("0%")).toBeInTheDocument();
  });

  test("should render basic Progress bar with correct percentage", () => {
    const { container, getByText } = render(<ProgressBar value={40} />);
    expect(container).toMatchSnapshot();
    expect(getByText("40%")).toBeInTheDocument();
  });

  test("should not go below zero", () => {
    const { container, getByText } = render(<ProgressBar value={-170} />);
    expect(container).toMatchSnapshot();
    expect(getByText("0%")).toBeInTheDocument();
  });

  test("should render exceeded limit with red background", () => {
    const { container, getByText } = render(
      <ProgressBar value={170} limit={100} />
    );
    expect(container).toMatchSnapshot();
    expect(getByText("170%")).toBeInTheDocument();
  });
});
