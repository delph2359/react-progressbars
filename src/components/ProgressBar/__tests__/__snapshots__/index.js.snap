// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress bar should not go below zero 1`] = `
<div>
  <div
    class="progress-bar"
    data-testid="progress-bar"
  >
    <div
      class="progress-fill"
      style="width: 0%;"
    />
    <span>
      0
      %
    </span>
  </div>
</div>
`;

exports[`Progress bar should render basic Progress bar (no props) 1`] = `
<div>
  <div
    class="progress-bar"
    data-testid="progress-bar"
  >
    <div
      class="progress-fill"
      style="width: 0%;"
    />
    <span>
      0
      %
    </span>
  </div>
</div>
`;

exports[`Progress bar should render basic Progress bar with correct percentage 1`] = `
<div>
  <div
    class="progress-bar"
    data-testid="progress-bar"
  >
    <div
      class="progress-fill"
      style="width: 40%;"
    />
    <span>
      40
      %
    </span>
  </div>
</div>
`;

exports[`Progress bar should render exceeded limit with red background 1`] = `
<div>
  <div
    class="progress-bar"
    data-testid="progress-bar"
  >
    <div
      class="progress-fill limit"
      style="width: 100%;"
    />
    <span>
      170
      %
    </span>
  </div>
</div>
`;
